File: crates/zoro-core/src/scan.rs
use std::fs;
use std::path::{Path, PathBuf};

use crate::model::{Block, ShellBlock, TagKind};
use crate::registry;

/// 扫描单个 Markdown 文件，按任意 `@<name>` 标签切分成内容块。
///
/// 返回块的 `library` 为空串，由 `scan_dir_named` / `Library::open` 填充。
pub fn scan_file(path: &Path) -> Result<Vec<Block>, ScanError> {
    let text = fs::read_to_string(path).map_err(ScanError::Io)?;
    Ok(scan_text(&text, path))
}

/// 对文本内容执行分块扫描（便于测试直接喂字符串）。
pub fn scan_text(text: &str, path: &Path) -> Vec<Block> {
    let mut scanner = Scanner::new(path);
    scanner.consume(text);
    scanner.blocks
}

/// 从 `start` 行开始，按统一边界现场截取一个块的原文。
///
/// 边界 = `start` 行 → 下一个任意 `@<name>` 标签行 / 文件末尾。
pub fn slice_block(text: &str, start: usize) -> String {
    let mut out: Vec<&str> = Vec::new();
    for (idx, line) in text.lines().enumerate() {
        let no = idx + 1;
        if no < start {
            continue;
        }
        if no > start && parse_directive(line).is_some() {
            break;
        }
        out.push(line);
    }
    out.join("\n")
}

#[derive(Debug)]
pub enum ScanError {
    Io(std::io::Error),
}

impl std::fmt::Display for ScanError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ScanError::Io(e) => write!(f, "io error: {e}"),
        }
    }
}

impl std::error::Error for ScanError {}

struct Scanner<'a> {
    path: &'a Path,
    blocks: Vec<Block>,
    last_heading: Option<String>,
    /// `last_heading` 是否仍然"紧贴"下一个标签（中间只允许空行 / 其他 `@` 行）。
    heading_fresh: bool,
    cur: Option<CurBlock>,
    /// 是否在等待 `@shell` 的紧随 fence。
    pending_shell: bool,
}

#[derive(Default)]
struct CurBlock {
    start: usize,
    title: Option<String>,
    kind: TagKind,
    terms: Vec<String>,
    lines: Vec<String>,
    shell: Option<ShellBlock>,
}

impl<'a> Scanner<'a> {
    fn new(path: &'a Path) -> Self {
        Scanner {
            path,
            blocks: Vec::new(),
            last_heading: None,
            heading_fresh: false,
            cur: None,
            pending_shell: false,
        }
    }

    fn consume(&mut self, text: &str) {
        // 正在收集的 `@shell` fence 状态。
        let mut fence_kind: Option<char> = None;
        let mut shell_lang: Option<String> = None;
        let mut shell_lines: Vec<usize> = Vec::new();

        for (idx, line) in text.lines().enumerate() {
            let line_no = idx + 1;

            // 0) 正在收集 shell fence：fence 内不解析标签。
            if let Some(kind) = fence_kind {
                if is_fence_close(line, kind) {
                    if let Some(cur) = self.cur.as_mut() {
                        cur.lines.push(line.to_string());
                        cur.shell = Some(ShellBlock {
                            lang: shell_lang.take(),
                            lines: std::mem::take(&mut shell_lines),
                        });
                    }
                    self.pending_shell = false;
                    fence_kind = None;
                } else {
                    shell_lines.push(line_no);
                    if let Some(cur) = self.cur.as_mut() {
                        cur.lines.push(line.to_string());
                    }
                }
                continue;
            }

            // 1) 标题：只记录最近标题，不参与边界。
            if let Some(rest) = line.strip_prefix("## ") {
                self.last_heading = Some(rest.trim().to_string());
                self.heading_fresh = true;
                continue;
            }
            if line.trim() == "##" {
                self.last_heading = Some(String::new());
                self.heading_fresh = true;
                continue;
            }

            // 2) 任意 `@<name>` 标签行：结束上一块，开新块。
            if let Some((name, value)) = parse_directive(line) {
                self.flush_current();
                let kind = registry::classify_tag(&name);
                let title = if self.heading_fresh {
                    self.last_heading.clone().filter(|h| !h.is_empty())
                } else {
                    None
                };
                self.pending_shell = kind.is_shell();
                self.cur = Some(CurBlock {
                    start: line_no,
                    title,
                    kind,
                    terms: split_ws(&value),
                    lines: vec![line.to_string()],
                    shell: None,
                });
                // `@` 行不破坏标题紧贴性。
                continue;
            }

            // 3) `@shell` 之后：开始收集紧随的 fence。
            if self.pending_shell {
                if let Some(kind) = fence_kind_of(line) {
                    fence_kind = Some(kind);
                    shell_lang = fence_lang(line);
                    shell_lines.clear();
                    if let Some(cur) = self.cur.as_mut() {
                        cur.lines.push(line.to_string());
                    }
                    // fence 正文会破坏标题紧贴性。
                    self.heading_fresh = false;
                    continue;
                }
            }

            // 4) 其余：正文。
            if !line.trim().is_empty() {
                self.heading_fresh = false;
            }
            if let Some(cur) = self.cur.as_mut() {
                cur.lines.push(line.to_string());
            }
        }

        // 文件结尾：未闭合的 shell fence 也照常绑定。
        if self.pending_shell {
            if let Some(cur) = self.cur.as_mut() {
                if cur.shell.is_none() && !shell_lines.is_empty() {
                    cur.shell = Some(ShellBlock {
                        lang: shell_lang.take(),
                        lines: std::mem::take(&mut shell_lines),
                    });
                }
            }
        }

        self.flush_current();
    }

    fn flush_current(&mut self) {
        if let Some(mut cur) = self.cur.take() {
            let title = cur
                .title
                .clone()
                .filter(|h| !h.is_empty())
                .unwrap_or_else(|| {
                    cur.terms
                        .first()
                        .cloned()
                        .unwrap_or_else(|| cur.kind.as_str())
                });
            self.blocks.push(Block {
                library: String::new(),
                title,
                kind: cur.kind,
                terms: std::mem::take(&mut cur.terms),
                path: self.path.to_path_buf(),
                start: cur.start,
                raw: cur.lines.join("\n"),
                shell: cur.shell,
            });
        }
    }
}

/// 解析顶格 `@<name> value`，返回 `(name, value)`；不匹配返回 None。
fn parse_directive(line: &str) -> Option<(String, String)> {
    let rest = line.strip_prefix('@')?;
    let name: String = rest
        .chars()
        .take_while(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_')
        .collect();
    if name.is_empty() {
        return None;
    }
    let after = &rest[name.len()..];
    if after.is_empty() || after.starts_with(' ') || after.starts_with('\t') {
        Some((name, after.trim().to_string()))
    } else {
        None
    }
}

fn split_ws(s: &str) -> Vec<String> {
    s.split_whitespace().map(|t| t.to_string()).collect()
}

/// fence 起始判定，返回定界符：'`' 或 '~'。
fn fence_kind_of(line: &str) -> Option<char> {
    let t = line.trim_start();
    if t.starts_with("```") {
        Some('`')
    } else if t.starts_with("~~~") {
        Some('~')
    } else {
        None
    }
}

fn fence_lang(line: &str) -> Option<String> {
    let t = line.trim_start();
    let rest = t.strip_prefix("```").or_else(|| t.strip_prefix("~~~"))?;
    let lang = rest.trim().split_whitespace().next().unwrap_or("");
    if lang.is_empty() {
        None
    } else {
        Some(lang.to_string())
    }
}

fn is_fence_close(line: &str, kind: char) -> bool {
    let t = line.trim_start();
    match kind {
        '`' => t.starts_with("```"),
        '~' => t.starts_with("~~~"),
        _ => false,
    }
}

/// 扫描一个内容根目录下所有 `*.md` / `*.mdx`，返回按路径稳定排序的块。
pub fn scan_dir(root: &Path) -> Result<Vec<Block>, ScanError> {
    scan_dir_inner(root, "")
}

/// 同 [`scan_dir`]，但把每个块的 `library` 置为给定库名。
pub fn scan_dir_named(root: &Path, library: &str) -> Result<Vec<Block>, ScanError> {
    scan_dir_inner(root, library)
}

fn scan_dir_inner(root: &Path, library: &str) -> Result<Vec<Block>, ScanError> {
    let mut files = Vec::new();
    collect_md(root, &mut files).map_err(ScanError::Io)?;
    files.sort();

    let mut blocks = Vec::new();
    for file in files {
        let rel = file.strip_prefix(root).unwrap_or(&file).to_path_buf();
        for mut b in scan_file(&file)? {
            b.path = rel.clone();
            b.library = library.to_string();
            blocks.push(b);
        }
    }
    Ok(blocks)
}

fn collect_md(root: &Path, out: &mut Vec<PathBuf>) -> std::io::Result<()> {
    if !root.is_dir() {
        return Ok(());
    }
    for entry in fs::read_dir(root)? {
        let entry = entry?;
        let path = entry.path();
        if entry.file_type()?.is_dir() {
            let name = entry.file_name();
            let name = name.to_string_lossy().to_string();
            if name.starts_with('.') {
                continue;
            }
            collect_md(&path, out)?;
        } else if let Some(ext) = path.extension().and_then(|e| e.to_str()) {
            if ext.eq_ignore_ascii_case("md") || ext.eq_ignore_ascii_case("mdx") {
                out.push(path);
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn splits_two_index_blocks() {
        let text = "\
## Head
@index alpha beta
body one
@index gamma
body two
";
        let blocks = scan_text(text, Path::new("a.md"));
        assert_eq!(blocks.len(), 2);
        assert_eq!(blocks[0].title, "Head");
        assert_eq!(blocks[0].terms, vec!["alpha", "beta"]);
        assert_eq!(blocks[0].start, 2);
        assert_eq!(blocks[1].title, "gamma");
        assert_eq!(blocks[1].start, 4);
    }

    #[test]
    fn shared_heading_when_adjacent() {
        let text = "\
## Head
@index alpha
body
@index beta
";
        let blocks = scan_text(text, Path::new("a.md"));
        assert_eq!(blocks.len(), 2);
        assert_eq!(blocks[0].title, "Head");
        // 第二个 @index 上方隔了正文，标题不再继承，退回搜索词首词
        assert_eq!(blocks[1].title, "beta");
    }

    #[test]
    fn binds_shell_fence() {
        let text = "\
@index run stuff
@shell
```bash
echo hi
echo bye
```
tail
";
        let blocks = scan_text(text, Path::new("a.md"));
        assert_eq!(blocks.len(), 2);
        let b = &blocks[1];
        assert_eq!(b.kind, TagKind::Shell);
        let shell = b.shell.as_ref().expect("shell payload");
        assert_eq!(shell.lang.as_deref(), Some("bash"));
        assert_eq!(shell.lines, vec![4, 5]);
        assert!(b.raw.contains("echo hi"));
    }

    #[test]
    fn any_tag_ends_previous_block() {
        let text = "\
@index alpha
index body
@shell git 丢弃
```bash
git checkout -- .
```
after
";
        let blocks = scan_text(text, Path::new("a.md"));
        assert_eq!(blocks.len(), 2);
        assert_eq!(blocks[0].kind, TagKind::Index);
        assert!(blocks[0].raw.contains("index body"));
        assert!(!blocks[0].raw.contains("@shell"));
        assert_eq!(blocks[1].kind, TagKind::Shell);
        assert_eq!(blocks[1].terms, vec!["git", "丢弃"]);
        assert!(blocks[1].raw.contains("after"));
    }

    #[test]
    fn unknown_tag_degrades_to_note() {
        let text = "\
@index alpha
body
@foo bar baz
payload
";
        let blocks = scan_text(text, Path::new("a.md"));
        assert_eq!(blocks.len(), 2);
        assert_eq!(blocks[1].kind, TagKind::Unknown("foo".to_string()));
        assert_eq!(blocks[1].terms, vec!["bar", "baz"]);
        assert!(blocks[1].raw.contains("payload"));
    }

    #[test]
    fn block_ends_at_eof() {
        let text = "@index only\nline one\nline two";
        let blocks = scan_text(text, Path::new("a.md"));
        assert_eq!(blocks.len(), 1);
        assert_eq!(blocks[0].raw.lines().count(), 3);
    }

    #[test]
    fn slice_block_matches_scanner_boundary() {
        let text = "\
## Head
@index alpha beta
line one
@shell git
```bash
echo hi
```
";
        // start = 2（@index alpha beta 行），到下一个标签（第 4 行）前结束
        assert_eq!(slice_block(text, 2), "@index alpha beta\nline one");
        // shell 块到 EOF
        let sliced = slice_block(text, 4);
        assert!(sliced.starts_with("@shell git"));
        assert!(sliced.ends_with("```"));
    }
}

File: core/scan.go
package core

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
	"unicode"
)

// ScanFile scans a single Markdown file, splitting it into content blocks.
// Returned blocks have an empty Library field, filled by ScanDirNamed / Library.Open.
func ScanFile(path string) ([]Block, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return ScanText(string(data), path), nil
}

// ScanText scans text content into blocks (convenient for testing with inline strings).
func ScanText(text, path string) []Block {
	s := newScanner(path)
	s.consume(text)
	return s.blocks
}

// SliceBlock cuts a block's raw text starting at the 1-based `start` line.
// Boundary = start line -> next `@<name>` tag line / EOF (see blockEnd).
//
// ⚠️ 标题行被跳过，与扫描器**逐字节一致**（consume 第 1 条不把标题行放进 Raw）：
// 一个块末尾的 `## 标题` 是**下一个**块的标题，不属于本块。少了这一条，
// LoadRaw 看到的范围就比 Block.Raw 大，改/删会顺手吃掉下一个块的标题。
func SliceBlock(text string, start int) string {
	lines := splitLines(text)
	from := start - 1
	if from < 0 || from >= len(lines) {
		return ""
	}
	out := make([]string, 0, blockEnd(lines, from)-from)
	for _, line := range lines[from:blockEnd(lines, from)] {
		if isHeadingLine(line) {
			continue
		}
		out = append(out, line)
	}
	return strings.Join(out, "\n")
}

// isHeadingLine 报告一行是否是扫描器眼里的**块标题行**：`## ` 开头，或整行只有 `##`。
// 标题归属只有这一处定义（scanner.consume 第 1 条、SliceBlock、write.go 的范围计算共用）。
// ⚠️ 缩进的 "  ## x" 不是标题（扫描器要求 ## 在第 0 列），它是正文。
func isHeadingLine(line string) bool {
	return strings.HasPrefix(line, "## ") || strings.TrimSpace(line) == "##"
}

// blockEnd returns the exclusive end of the block whose tag line is `lines[from]`.
//
// ⚠️ 这是块边界的**唯一定义处**，与 scanner.consume 逐条对齐；SliceBlock（预览 / Raw）
// 和 write.go 的改 / 删都走它。改这里等于同时改「用户看到什么」和「会删掉什么」，
// 两者必须始终一致（write.go 的 L2）。
//
//	扫描器第 0 条 → 代码围栏内部的 `@` 行是代码，不是边界；
//	扫描器第 2 条 → 第 0 列的 `@<name>` 行结束本块（缩进的 @ 行不算，parseDirective 要求第 0 列）；
//	扫描器第 3 条 → 只有紧跟 `@shell` 的**第一个**围栏受保护，围栏闭合后 pendingShell 清零。
//
// 少了围栏这一条，一个含 `@echo off` 的 bat 代码块会被从中间劈开 ——
// 预览只显示半截，删除会留下一个不闭合的围栏，文件直接坏掉。
func blockEnd(lines []string, from int) int {
	pendingShell := false
	if name, _, ok := parseDirective(lines[from]); ok {
		pendingShell = ClassifyTag(name).IsShell()
	}
	var fence rune
	for i := from + 1; i < len(lines); i++ {
		line := lines[i]
		if fence != 0 {
			if isFenceClose(line, fence) {
				fence, pendingShell = 0, false
			}
			continue
		}
		if _, _, ok := parseDirective(line); ok {
			return i
		}
		if pendingShell {
			if k := fenceKindOf(line); k != 0 {
				fence = k
			}
		}
	}
	return len(lines)
}

type scanner struct {
	path         string
	blocks       []Block
	lastHeading  string
	headingFresh bool
	cur          *curBlock
	pendingShell bool
}

type curBlock struct {
	start int
	title string
	kind  TagKind
	terms []string
	lines []string
	shell *ShellBlock
}

func newScanner(path string) *scanner {
	return &scanner{path: path}
}

func (s *scanner) consume(text string) {
	var (
		fenceKind  rune
		shellLang  string
		shellLines []int
	)

	for idx, line := range splitLines(text) {
		lineNo := idx + 1

		// 0) Inside a shell fence: `@` lines are code, never tags.
		if fenceKind != 0 {
			if isFenceClose(line, fenceKind) {
				if s.cur != nil {
					s.cur.lines = append(s.cur.lines, line)
					s.cur.shell = &ShellBlock{Lang: shellLang, Lines: shellLines}
				}
				s.pendingShell = false
				fenceKind = 0
			} else {
				shellLines = append(shellLines, lineNo)
				if s.cur != nil {
					s.cur.lines = append(s.cur.lines, line)
				}
			}
			continue
		}

		// 1) Heading: record the latest heading only; it is never a boundary.
		//    ⚠️ 标题行**不进 cur.lines**（所以也不在 Block.Raw 里），它属于下一个块的标题。
		//    判据只有 isHeadingLine 一处定义，SliceBlock / write.go 与这里必须一致。
		if isHeadingLine(line) {
			if rest, ok := strings.CutPrefix(line, "## "); ok {
				s.lastHeading = strings.TrimSpace(rest)
			} else {
				s.lastHeading = ""
			}
			s.headingFresh = true
			continue
		}

		// 2) Any `@<name>` tag line ends the previous block and starts a new one.
		if name, value, ok := parseDirective(line); ok {
			s.flushCurrent()
			kind := ClassifyTag(name)
			title := ""
			if s.headingFresh {
				title = s.lastHeading
			}
			s.pendingShell = kind.IsShell()
			s.cur = &curBlock{
				start: lineNo,
				title: title,
				kind:  kind,
				terms: strings.Fields(value),
				lines: []string{line},
			}
			// A tag line does not break heading freshness.
			continue
		}

		// 3) Immediately after `@shell`: start collecting the following fence.
		if s.pendingShell {
			if k := fenceKindOf(line); k != 0 {
				fenceKind = k
				shellLang = fenceLang(line)
				shellLines = nil
				if s.cur != nil {
					s.cur.lines = append(s.cur.lines, line)
				}
				s.headingFresh = false
				continue
			}
		}

		// 4) Everything else is body text.
		if strings.TrimSpace(line) != "" {
			s.headingFresh = false
		}
		if s.cur != nil {
			s.cur.lines = append(s.cur.lines, line)
		}
	}

	// Unclosed shell fence at EOF is still bound to the block.
	if s.pendingShell {
		if s.cur != nil && s.cur.shell == nil && len(shellLines) > 0 {
			s.cur.shell = &ShellBlock{Lang: shellLang, Lines: shellLines}
		}
	}

	s.flushCurrent()
}

func (s *scanner) flushCurrent() {
	if s.cur == nil {
		return
	}
	cur := s.cur
	s.cur = nil

	title := cur.title
	if title == "" {
		if len(cur.terms) > 0 {
			title = cur.terms[0]
		} else {
			title = cur.kind.String()
		}
	}
	s.blocks = append(s.blocks, Block{
		Library: "",
		Title:   title,
		Kind:    cur.kind,
		Terms:   cur.terms,
		Path:    s.path,
		Start:   cur.start,
		Raw:     strings.Join(cur.lines, "\n"),
		Shell:   cur.shell,
	})
}

// parseDirective parses a top-level `@name value` line; returns (name, value, ok).
func parseDirective(line string) (string, string, bool) {
	if !strings.HasPrefix(line, "@") {
		return "", "", false
	}
	rest := line[1:]
	i := 0
	for i < len(rest) {
		c := rest[i]
		if isASCIIName(c) {
			i++
			continue
		}
		break
	}
	if i == 0 {
		return "", "", false
	}
	name := rest[:i]
	after := rest[i:]
	if after == "" || after[0] == ' ' || after[0] == '\t' {
		return name, strings.TrimSpace(after), true
	}
	return "", "", false
}

func isASCIIName(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-' || c == '_'
}

func fenceKindOf(line string) rune {
	t := trimLeftSpace(line)
	if strings.HasPrefix(t, "```") {
		return '`'
	}
	if strings.HasPrefix(t, "~~~") {
		return '~'
	}
	return 0
}

func fenceLang(line string) string {
	t := trimLeftSpace(line)
	rest := ""
	for _, marker := range []string{"```", "~~~"} {
		if strings.HasPrefix(t, marker) {
			rest = strings.TrimPrefix(t, marker)
			break
		}
	}
	fields := strings.Fields(rest)
	if len(fields) == 0 {
		return ""
	}
	return fields[0]
}

func isFenceClose(line string, kind rune) bool {
	t := trimLeftSpace(line)
	switch kind {
	case '`':
		return strings.HasPrefix(t, "```")
	case '~':
		return strings.HasPrefix(t, "~~~")
	}
	return false
}

func trimLeftSpace(s string) string {
	return strings.TrimLeftFunc(s, unicode.IsSpace)
}

// mdFile is one content file candidate for scanning / fingerprinting.
type mdFile struct {
	Rel   string // library-relative slash path
	Abs   string // absolute filesystem path
	Size  int64
	MTime time.Time
}

// ScanDir scans all `*.md` / `*.mdx` under a content root, sorted by path.
func ScanDir(root string) ([]Block, error) {
	return scanDirInner(root, "")
}

// ScanDirNamed is like ScanDir but sets each block's Library to `library`.
func ScanDirNamed(root, library string) ([]Block, error) {
	return scanDirInner(root, library)
}

func scanDirInner(root, library string) ([]Block, error) {
	files, err := listMDFiles(root)
	if err != nil {
		return nil, err
	}

	blocks := make([]Block, 0, len(files)*4)
	for _, file := range files {
		bs, err := ScanFile(file.Abs)
		if err != nil {
			return nil, err
		}
		for i := range bs {
			bs[i].Path = filepath.ToSlash(file.Rel)
			bs[i].Library = library
			blocks = append(blocks, bs[i])
		}
	}
	return blocks, nil
}

// listMDFiles walks the content root and returns all markdown files sorted by
// relative path, skipping hidden directories (`.git / .zoro / ...`).
func listMDFiles(root string) ([]mdFile, error) {
	var files []mdFile
	err := filepath.WalkDir(root, func(p string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			if p != root && strings.HasPrefix(d.Name(), ".") {
				return filepath.SkipDir
			}
			return nil
		}
		ext := strings.ToLower(filepath.Ext(d.Name()))
		if ext != ".md" && ext != ".mdx" {
			return nil
		}
		info, err := d.Info()
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(root, p)
		if err != nil {
			rel = p
		}
		files = append(files, mdFile{
			Rel:   filepath.ToSlash(rel),
			Abs:   p,
			Size:  info.Size(),
			MTime: info.ModTime(),
		})
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Slice(files, func(i, j int) bool { return files[i].Rel < files[j].Rel })
	return files, nil
}

// splitLines mimics Rust str::lines(): \r\n -> \n, and no trailing empty line.
func splitLines(text string) []string {
	text = strings.ReplaceAll(text, "\r\n", "\n")
	lines := strings.Split(text, "\n")
	if n := len(lines); n > 0 && lines[n-1] == "" {
		lines = lines[:n-1]
	}
	for i := range lines {
		lines[i] = strings.TrimSuffix(lines[i], "\r")
	}
	return lines
}
